File: 199_binary_tree_right_side_view/rightSideView.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function rightSideView(root: TreeNode | null): number[] {
  const res = [];

  dfs(root, 0, res);

  return res;
}

function dfs(root: TreeNode | null, level: number, res: number[]) {
  if (root) {
    if (res.length <= level) res.push(root.val);

    res[level] = root.val;

    dfs(root.left, level + 1, res);
    dfs(root.right, level + 1, res);
  }
}

File: 199_binary_tree_right_side_view/rightSideView.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func rightSideView(root *TreeNode) []int {
    res := make([]int, 0)

    dfs(root, 0, &res)

    return res
}

func dfs(root *TreeNode, level int, res *[]int) {
    if root != nil {
        if len(*res) <= level {
            *res = append(*res, root.Val)
        }

        (*res)[level] = root.Val

        dfs(root.Left, level + 1, res)
        dfs(root.Right, level + 1, res)
    }
}